### PR TITLE
Issue #78 Implementing Number Format Parser

### PR DESCRIPTION
**Pull request by robons**
This PR gets basic Unicode [UTS-35/TR-35/UAX-35](http://www.unicode.org/reports/tr35/tr35-numbers.html#Number_Format_Patterns) number format validation functioning within the standards set out by unicode and the  W3C CSV-W working group. Notably, all [W3C CSV-W tests](https://w3c.github.io/csvw/tests/) relating to number format validation now pass successfully.

N.B. It is not required that we implement the full UTS-35 specification. This PR does support the [padding](http://www.unicode.org/reports/tr35/tr35-numbers.html#Padding), [rounding](http://www.unicode.org/reports/tr35/tr35-numbers.html#Rounding) or [significant digits](http://www.unicode.org/reports/tr35/tr35-numbers.html#sigdig) functionality, however it is entirely possible to retrofit this functionality at a later point in time, if desired.

> Implementations MUST recognise number format patterns containing the symbols 0, #, the specified decimalChar (or "." if unspecified), the specified groupChar (or "," if unspecified), E, +, % and ‰. Implementations MAY additionally recognise number format patterns containing other special pattern characters defined in [UAX35]
[W3C CSV-W - Number Format Minimum Standards](https://www.w3.org/TR/2015/REC-tabular-data-model-20151217/#formats-for-numeric-types)

N.B. This approach parses all numbers to the `BigDecimal` infinite precision datatype. This isn't likely to be the most performant approach, but it is simple to convert the parsed value from a string into a numeric representation.

Newly created issues: #98, #99

This PR satisfies #90.

### CODE REVIEW

File: src/main/scala/numberformatparser/NumberFormatParser.scala [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+                isFractionalPart
+              ) ^^ ensureCorrectNumDigits,
+              digitAndGroupChars
+            ) match {
+              case Success(digitsWithoutGroupingChar, _) =>
+                Success(digitsWithoutGroupingChar, currentPosition)
+              case Failure(err, _) => Failure(err, currentPosition)
+            }
+          } catch {
+            case e => Failure(e.getMessage, currentPosition)
```
Comment (robons) [inline src/main/scala/numberformatparser/NumberFormatParser.scala:228]: todo: log debug?

File: src/main/scala/Column.scala [hunk 2]
```diff
@@ -464,12 +465,19 @@ case class Column private (
...
+  lazy val numberParserForFormat: Either[String, NumberParser] =
+    try {
+      Right(
+        LdmlNumberFormatParser(
+          format.flatMap(f => f.groupChar).getOrElse(','),
+          format.flatMap(f => f.decimalChar).getOrElse('.')
+        ).getParserForFormat(format.flatMap(f => f.pattern).get)
+      )
+    } catch {
+      case e =>
```
[the remark this answers is not in the document]
Comment (cjfairchild) [reply src/main/scala/Column.scala:477]: Having seen the latter code for debug extensions - I'm not sure the caught exception will be logged at all as there's a conditional as ```debug(error: Throwable)``` in the logging code.

**Review by cjfairchild — CHANGES_REQUESTED** [review]
Please see comments for actionable issues - they're minor - one for error catching and likely compiler complaints and another for documentation around the ```ArrayCursor``` class.

File: src/main/scala/Column.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (robons) [reply src/main/scala/Column.scala:477]: > Throwing unspecified e is generally frowned upon
Indeed, I'm just not ready to invest the time necessary to remove all the general catches everywhere right now.

> one could use 'case e : Throwable => ...'
Will do

File: src/main/scala/models/ArrayCursor.scala [hunk 3]
```diff
@@ -0,0 +1,42 @@
+package models
+
+import scala.reflect.ClassTag
+
+case class ArrayCursor[T: ClassTag](private val values: Seq[T]) {
```
[the remark this answers is not in the document]
Comment (robons) [reply src/main/scala/models/ArrayCursor.scala:5]: Good call

**Pull request merged**